exams/api.py

```diff
@@ -32,3 +57,115 @@ def sso_digest(client_candidate_id, timestamp, session_timeout):
         str(client_candidate_id),
     ]).encode('iso-8859-1')
     return hashlib.sha256(data).hexdigest()
+
+
+def authorize_for_exam_run(mmtrack, course_run, exam_run):
+    """
+    Authorize user for exam if he has paid for course and passed course.
+
+    Args:
+        mmtrack (dashboard.utils.MMTrack): a instance of all user information about a program.
+        course_run (courses.models.CourseRun): A CourseRun object.
+        exam_run (exams.models.ExamRun): the ExamRun we're authorizing for
+    """
+    if not exam_run.is_schedulable:
+        raise ExamAuthorizationException("Exam isn't schedulable currently: {}".format(exam_run))
+
+    # If user has not paid for course then we dont need to process authorization
+    if not mmtrack.has_paid(course_run.edx_course_key):
+        errors_message = MESSAGE_NOT_ELIGIBLE_TEMPLATE.format(
+            user=mmtrack.user.username,
+            course_id=course_run.edx_course_key
+        )
+        raise ExamAuthorizationException(errors_message)
+
+    # if user paid for a course then create his exam profile if it is not creaated yet
+    ExamProfile.objects.get_or_create(profile=mmtrack.user.profile)
+
+    # if they didn't pass, they don't get authorized
+    if not mmtrack.has_passed_course(course_run.edx_course_key):
+        errors_message = MESSAGE_NOT_PASSED_OR_EXIST_TEMPLATE.format(
+            user=mmtrack.user.username,
+            course_id=course_run.edx_course_key
+        )
+        raise ExamAuthorizationException(errors_message)
+
+    ExamAuthorization.objects.get_or_create(
+        user=mmtrack.user,
+        course=course_run.course,
+        exam_run=exam_run,
+        defaults={  # write these out even though they're deprecated as putting nulls in here forfeits a db rollback
+            'date_first_eligible': exam_run.date_first_eligible,
+            'date_last_eligible': exam_run.date_last_eligible,
+        }
+    )
+    log.info(
+        '[Exam authorization] user "%s" is authorized for the exam for course id "%s"',
+        mmtrack.user.username,
+        course_run.edx_course_key
+    )
+
+
+def authorize_for_latest_passed_course(mmtrack, exam_run):
+    """
+    Authorize user for exam on given course. Using latest passed run.
```

Comment:
This says 'latest passed run' but the queryset looks like it just gets the latest run. Is it assumed to have been passed if the `CachedEnrollment` exists?